dart/lib/src/reference.dart

```diff
@@ -0,0 +1,416 @@
+      }
+      if(ValueTypeUtils.isFixedTypedVector(_valueType)) {
+        flx._valueType = ValueTypeUtils.fixedTypedVectorElementType(_valueType);
+        return flx;
+      }
+      var packedType = _buffer.getUint8(_indirect + length * _byteWidth + index);
+      return Reference._(_buffer, elementOffset, BitWidthUtil.fromByteWidth(_byteWidth), packedType);
+    }
+    if (key is String && _valueType == ValueType.Map) {
+      var index = _keyIndex(key);
+      if (index != null) {
+        return _valueForIndex(index);
+      }
+    }
+    return null;
+  }
+
+  /// Get an iterable if the underlying flexBuffer value is a vector.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get vectorIterable {
+    if(isVector == false) {
+      throw Exception('Value is not a vector. It is: ${_valueType}');
+    }
+    return _VectorIterator(this);
+  }
+
+  /// Get an iterable for keys if the underlying flexBuffer value is a map.
+  /// Otherwise throws an exception.
+  Iterable<String> get mapKeyIterable {
+    if(isMap == false) {
+      throw Exception('Value is not a map. It is: ${_valueType}');
+    }
+    return _MapKeyIterator(this);
+  }
+
+  /// Get an iterable for values if the underlying flexBuffer value is a map.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get mapValueIterable {
+    if(isMap == false) {
+      throw Exception('Value is not a map. It is: ${_valueType}');
+    }
+    return _MapValueIterator(this);
+  }
+
+  /// Returns the length of the the underlying FlexBuffer value.
+  /// If the underlying value is [null] the length is 0.
+  /// If the underlying value is a number, or a bool, the length is 1.
+  /// If the underlying value is a vector, or map, the length reflects number of elements / element pairs.
+  /// If the values is a string or a blob, the length reflects a number of bytes the value occupies (strings are encoded in utf8 format).
+  int get length {
+    if (_length != null) {
+      return _length;
+    }
+    // needs to be checked before more generic isAVector
+    if(ValueTypeUtils.isFixedTypedVector(_valueType)) {
+      _length = ValueTypeUtils.fixedTypedVectorElementSize(_valueType);
+    } else if(_valueType == ValueType.Blob || ValueTypeUtils.isAVector(_valueType) || _valueType == ValueType.Map){
+      _length = _readInt(_indirect - _byteWidth, BitWidthUtil.fromByteWidth(_byteWidth));
+    } else if (_valueType == ValueType.Null) {
+      _length = 0;
+    } else if (_valueType == ValueType.String) {
+      var indirect = _indirect;
+      var size_byte_width = _byteWidth;
+      var size = _readInt(indirect - size_byte_width, BitWidthUtil.fromByteWidth(size_byte_width));
+      while (_buffer.getInt8(indirect + size) != 0) {
+        size_byte_width <<= 1;
+        size = _readInt(indirect - size_byte_width, BitWidthUtil.fromByteWidth(size_byte_width));
+      }
+      _length = size;
+    } else if (_valueType == ValueType.Key) {
+      var indirect = _indirect;
```

Comment:
unused?

Reply:
There are multiple comments regarding `_indirect` I will comment here and resolve the others. `_indirect` is a computed property, when it is used more than once in a computation branch I cache it into a local variable. So it should not be computed more than once.
We could cache the computation on a class level, however that would mean that we have to check for cached value every time, which would be sub optimal specifically when the variable is used in a loop. Also as the `Reference` is polymorphic, is seems like we using the `_indirect` very often, but it is actually not the case. We use it mostly once per value type. So introduction of a cached property is IMHO not justified. It would make the code read better, sacrificing a bit of performance.

Reply:
In that case, can we add some comments in the code to this effect?  Perhaps on the `_indirect` getter as a doc comment.

E.g.

```
/// Computes the blah blah value.
///
/// To optimize for the more common case of being called only once, this
/// value is not cached. Callers that need to use it more than once should
/// cache the return value in a local variable.
```